modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.painless.antlr;
+
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.LexerNoViableAltException;
+import org.antlr.v4.runtime.Token;
+import org.elasticsearch.painless.lookup.PainlessLookup;
+
+/**
+ * A lexer that is customized for painless. It:
+ * <ul>
+ * <li>Overrides the default error behavior to fail on the first error.
+ * <li>Stores the last token in case we need to do lookbehind for semicolon insertion and regex vs division detection.
+ * <li>Implements the regex vs division detection.
+ * <li>Insert semicolons where they'd improve the language's readability. Rather than hack this into the parser and create a ton of
+ * ambiguity we hack them here where we can use heuristics to do it quickly.
+ * <li>Enhances the error message when a string contains invalid escape sequences to include a list of valid escape sequences.
+ * </ul>
+ */
+public final class EnhancedSuggestLexer extends SuggestLexer {
+
+    private Token current = null;
+    private final PainlessLookup painlessLookup;
+
+    public EnhancedSuggestLexer(CharStream charStream, PainlessLookup painlessLookup) {
+        super(charStream);
+        this.painlessLookup = painlessLookup;
+    }
+
+    @Override
+    public Token nextToken() {
+        current = super.nextToken();
+        return current;
+    }
+
+    @Override
+    public void recover(final LexerNoViableAltException lnvae) {
+        if (this._mode != PainlessLexer.DEFAULT_MODE) {
+            this._mode = DEFAULT_MODE;
```

Comment:
Is this so that we can keep suggesting if we fail in `DOTID` or `DOTINTEGER`?

Reply:
Yes! We drop back to default mode so we can continue lexing.